components/experimental/src/duration/format.rs

```diff
@@ -147,39 +151,51 @@ impl<'a> FormattedDuration<'a> {
     /// Section 1.1.7
     /// Computes the sum of all values in durationFormat units with "fractional" style,
     /// expressed as a fraction of the smallest unit of durationFormat that does not use "fractional" style.
-    fn add_fractional_digits(&self, original: u64) -> FixedDecimal {
-        // convert to u128 to prevent overflow
-        let mut result: u128 = original.into();
-        let mut exponent = 0;
-
-        // 3. For each row of Table 2, except the header row, in table order, do
-        // a. Let style be the value of durationFormat's internal slot whose name is the Style Slot value of the current row.
-        for (style, value) in [
-            (self.fmt.options.millisecond, self.duration.milliseconds),
-            (self.fmt.options.microsecond, self.duration.microseconds),
-            (self.fmt.options.nanosecond, self.duration.nanoseconds),
-        ] {
-            // b. If style is "fractional", then
-            if style == FieldStyle::Fractional {
-                // i. Assert: The Unit value of the current row is "milliseconds", "microseconds", or "nanoseconds".
-                // ii. Let value be the value of duration's field whose name is the Value Field value of the current row.
-                // iii. Set value to value / (10 ^ exponent).
-                // iv. Set result to result + value.
-                result *= u128::pow(10, 3);
-                result += u128::from(value);
-                // v. Set exponent to exponent + 3.
-                exponent += 3;
+    /// then adds the first non-fractional unit and returns it.
+    ///
+    /// Divergence from standard: adds the first non-fractional unit as well.
+    fn add_fractional_digits(&self) -> FixedDecimal {
+        if self.fmt.options.nanosecond == FieldStyle::Fractional {
```

Comment:
Observation/Suggestion: I think you can reduce the nesting in this function a bit, but it is much better than it was yesterday.

Reply:
I've changed it to a loop.